pkg/controller/unifiedpushserver/unifiedpushserver_controller.go

```diff
@@ -197,9 +200,151 @@ func (r *ReconcileUnifiedPushServer) Reconcile(request reconcile.Request) (recon
+			return reconcile.Result{}, err
+		}
+
+		foundAddressSpace := &enmassev1beta.AddressSpace{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: addressSpace.Name, Namespace: addressSpace.Namespace}, foundAddressSpace)
+		if err != nil && errors.IsNotFound(err) {
+			reqLogger.Info("Creating a new Address Space", "AddressSpace.Namespace", addressSpace.Namespace, "AddressSpace.Name", addressSpace.Name)
+			err = r.client.Create(context.TODO(), addressSpace)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+		} else if err != nil {
+			return reconcile.Result{}, err
+		} else if !foundAddressSpace.Status.IsReady {
+			return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+		} else {
+			reqLogger.Info("Found AddressSpace for UPS")
+		}
+
+		/*Address space exists and is ready*/
+
+		for _, status := range foundAddressSpace.Status.EndpointStatus {
+			if status.Name == "messaging" { //magic value
+				addressSpaceURL := status.ServiceHost
+				addressConfigMap := newAMQConfigMap(instance, addressSpaceURL)
+				err = r.client.Create(context.TODO(), addressConfigMap)
+				if err != nil && !errors.IsAlreadyExists(err) {
+					return reconcile.Result{}, err
+				}
+
+			}
+		}
+
+		//check that user exists
+		user, err := newMessagingUser(instance)
+
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Set UnifiedPushServer instance as the owner and controller
+		if err := controllerutil.SetControllerReference(instance, user, r.scheme); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		foundUser := &messaginguserv1beta.MessagingUser{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: user.Name, Namespace: user.Namespace}, foundUser)
+		if err != nil && errors.IsNotFound(err) {
+			reqLogger.Info("Creating a new MessagingUser", "MessagingUser.Namespace", user.Namespace, "MessagingUser.Name", user.Name)
+			err = r.client.Create(context.TODO(), user)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			secret := newAMQSecret(instance, string(user.Spec.Authentication.Password))
+			err = r.client.Create(context.TODO(), secret)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			return reconcile.Result{RequeueAfter: time.Second * 1}, nil
+		} else if err != nil {
+			return reconcile.Result{}, err
+		} /*User exists and is ready*/
+
+		/*User exists and is ready*/
+		//check that addresses exist
+		//queues
+		queues := []string{"APNsPushMessageQueue", "APNsTokenBatchQueue", "GCMPushMessageQueue", "GCMTokenBatchQueue", "WNSPushMessageQueue", "WNSTokenBatchQueue", "MetricsQueue", "TriggerMetricCollectionQueue", "TriggerVariantMetricCollectionQueue", "BatchLoadedQueue", "AllBatchesLoadedQueue", "FreeServiceSlotQueue"}
+		requeueCreate := false
+		for _, address := range queues {
+			queue := newQueue(instance, address)
+			foundQueue := &enmassev1beta.Address{}
```

Comment:
I'm confused between "queue" and "topic" and "address". Are these all synonyms in enmasse? From my reading here, I'm guessing that "queue" and "topic" are types of "address". Is that right?

Reply:
They are both addresses with slightly different creation semantics that I wasn't comfortable abstracting out.